File: scala-aoc/src/main/scala/io/c6/aoc/y2022/Day02.scala
package io.c6.aoc.y2022

import io.c6.aoc.util.BaseSolution
import io.c6.aoc.util.Day.*
import io.c6.aoc.util.InputType.*
import io.c6.aoc.util.Year.*

import scala.language.implicitConversions

@main def _2022_02: Unit = Day02()

object Day02 extends BaseSolution(_2022, _02):
  import StringConversions.given

  private sealed trait Outcome(val points: Int)

  private object Win extends Outcome(6)

  private object Draw extends Outcome(3)

  private object Loose extends Outcome(0)

  private sealed trait Shape(val points: Int)

  private object Rock extends Shape(1)

  private object Paper extends Shape(2)

  private object Scissors extends Shape(3)

  private sealed trait OpponentPlay(val shape: Shape)

  private object A extends OpponentPlay(Rock)

  private object B extends OpponentPlay(Paper)

  private object C extends OpponentPlay(Scissors)

  private sealed trait MyResponse(val shape: Shape)

  private object X extends MyResponse(Rock)

  private object Y extends MyResponse(Paper)

  private object Z extends MyResponse(Scissors)

  private val outcomes: Map[(Shape, Shape), Outcome] = Map(
    (Rock, Rock) -> Draw,
    (Rock, Paper) -> Win,
    (Rock, Scissors) -> Loose,
    (Paper, Rock) -> Loose,
    (Paper, Paper) -> Draw,
    (Paper, Scissors) -> Win,
    (Scissors, Rock) -> Win,
    (Scissors, Paper) -> Loose,
    (Scissors, Scissors) -> Draw
  )

  private val myResponses: Map[(Shape, Outcome), Shape] = Map(
    (Rock, Draw) -> Rock,
    (Rock, Win) -> Paper,
    (Rock, Loose) -> Scissors,
    (Paper, Loose) -> Rock,
    (Paper, Draw) -> Paper,
    (Paper, Win) -> Scissors,
    (Scissors, Win) -> Rock,
    (Scissors, Loose) -> Paper,
    (Scissors, Draw) -> Scissors
  )

  private type PlayPair = (OpponentPlay, MyResponse)

  private object StringConversions:
    private given toOpponentPlay: Conversion[String, OpponentPlay] = str => str match
      case "A" => A
      case "B" => B
      case _ => C

    private given toMyResponse: Conversion[String, MyResponse] = str => str match
      case "X" => X
      case "Y" => Y
      case _ => Z

    given Conversion[String, PlayPair] = str => str.split(" ") match
      case Array(abc, xyz) => toOpponentPlay(abc) -> toMyResponse(xyz)

  private def pointsForStrategy1(playPair: PlayPair): Int = playPair match
    case abc -> xyz =>
      outcomes(abc.shape, xyz.shape).points + xyz.shape.points

  private def pointsForStrategy2(playPair: PlayPair): Int = playPair match
    case abc -> xyz => xyz match
      case X => myResponses(abc.shape, Loose).points + Loose.points
      case Y => myResponses(abc.shape, Draw).points + Draw.points
      case Z => myResponses(abc.shape, Win).points + Win.points

  protected override def part1Solution: Seq[String] => Unit = input =>
    val result = input.map(pointsForStrategy1(_)).sum
    println(s"Result for part 1: $result")

  protected override def part2Solution: Seq[String] => Unit = input =>
    val result = input.map(pointsForStrategy2(_)).sum
    println(s"Result for part 2: $result")

File: java-aoc/src/main/java/io/c6/aoc/y2022/Day02.java
package io.c6.aoc.y2022;

import io.c6.aoc.BaseSolution;
import io.c6.aoc.Pair;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;

import static io.c6.aoc.Day._02;
import static io.c6.aoc.Year._2022;
import static io.c6.aoc.y2022.Day02.Outcome.*;
import static io.c6.aoc.y2022.Day02.Shape.*;

public final class Day02 extends BaseSolution {

	private static final Logger LOGGER = LoggerFactory.getLogger(Day02.class);

	private static final Map<Pair<Shape, Shape>, Outcome> outcomes = Map.of(
			Pair.of(Rock, Rock), Draw,
			Pair.of(Rock, Paper), Win,
			Pair.of(Rock, Scissors), Loose,
			Pair.of(Paper, Rock), Loose,
			Pair.of(Paper, Paper), Draw,
			Pair.of(Paper, Scissors), Win,
			Pair.of(Scissors, Rock), Win,
			Pair.of(Scissors, Paper), Loose,
			Pair.of(Scissors, Scissors), Draw
	);

	private static final Map<Pair<Shape, Outcome>, Shape> myResponses = Map.of(
			Pair.of(Rock, Draw), Rock,
			Pair.of(Rock, Win), Paper,
			Pair.of(Rock, Loose), Scissors,
			Pair.of(Paper, Loose), Rock,
			Pair.of(Paper, Draw), Paper,
			Pair.of(Paper, Win), Scissors,
			Pair.of(Scissors, Win), Rock,
			Pair.of(Scissors, Loose), Paper,
			Pair.of(Scissors, Draw), Scissors
	);

	public Day02() {
		super(_2022, _02);
	}

	public static void main(final String... args) {
		new Day02().run();
	}

	@Override
	public void solution() {
		final var result1 = getInput()
				.map(s -> s.split(" "))
				.map(arr -> Pair.of(OpponentPlay.valueOf(arr[0]), MyResponse.valueOf(arr[1])))
				.map(this::pointsForStrategy1)
				.reduce(0, Integer::sum);
		LOGGER.info("Total score for given first strategy: {}", result1);

		final var result2 = getInput()
				.map(s -> s.split(" "))
				.map(arr -> Pair.of(OpponentPlay.valueOf(arr[0]), MyResponse.valueOf(arr[1])))
				.map(this::pointsForStrategy2)
				.reduce(0, Integer::sum);
		LOGGER.info("Total score for given second strategy: {}", result2);
	}

	private int pointsForStrategy1(final Pair<OpponentPlay, MyResponse> pair) {
		final var shapes = pair.map(l -> l.shape, r -> r.shape);
		return outcomes.get(shapes).points + shapes.right.points;
	}

	private int pointsForStrategy2(final Pair<OpponentPlay, MyResponse> pair) {
		switch (pair.right) {
			case X: return myResponses.get(pair.map(l -> l.shape, r -> Loose)).points + Loose.points;
			case Y: return myResponses.get(pair.map(l -> l.shape, r -> Draw)).points + Draw.points;
			case Z: return myResponses.get(pair.map(l -> l.shape, r -> Win)).points + Win.points;
			default: return Loose.points;
		}
	}

	enum Outcome {
		Win(6), Draw(3), Loose(0);

		final int points;

		Outcome(final int points) {
			this.points = points;
		}
	}

	enum Shape {
		Rock(1), Paper(2), Scissors(3);

		final int points;

		Shape(final int points) {
			this.points = points;
		}
	}

	enum OpponentPlay {
		A(Rock), B(Paper), C(Scissors);

		final Shape shape;

		OpponentPlay(final Shape shape) {
			this.shape = shape;
		}
	}

	enum MyResponse {
		X(Rock), Y(Paper), Z(Scissors);

		final Shape shape;

		MyResponse(final Shape shape) {
			this.shape = shape;
		}
	}
}
